src/core/qgsmaprendererjob.cpp

```diff
@@ -363,6 +385,177 @@ LayerRenderJobs QgsMapRendererJob::prepareJobs( QPainter *painter, QgsLabelingEn
   return layerJobs;
 }
 
+LayerRenderJobs QgsMapRendererJob::prepareSecondPassJobs( LayerRenderJobs &firstPassJobs, LabelRenderJob &labelJob )
```

Comment:
this is a very long function with a lot of logic - would be great to split it into smaller chunks to make it easier to understand